zcash_client_sqlite/src/wallet/commitment_tree.rs

```diff
@@ -746,15 +757,98 @@ pub(crate) fn truncate_checkpoints(
     Ok(())
 }
 
+pub(crate) fn put_shard_roots<
+    H: Hashable + HashSer + Clone + Eq,
+    const DEPTH: u8,
+    const SHARD_HEIGHT: u8,
+>(
+    conn: &rusqlite::Transaction<'_>,
+    table_prefix: &'static str,
+    start_index: u64,
+    roots: &[CommitmentTreeRoot<H>],
+) -> Result<(), ShardTreeError<Error>> {
+    if roots.is_empty() {
+        // nothing to do
+        return Ok(());
+    }
+
+    // We treat the cap as a DEPTH-SHARD_HEIGHT tree so that we can make a batch insertion of
+    // root data using `Position::from(start_index)` as the starting position and treating the
+    // roots as level-0 leaves.
+    let cap = LocatedTree::from_parts(
+        Address::from_parts((DEPTH - SHARD_HEIGHT).into(), 0),
+        get_cap(conn, table_prefix).map_err(ShardTreeError::Storage)?,
+    );
+
+    let cap_result = cap
+        .batch_insert(
+            Position::from(start_index),
+            roots.iter().map(|r| {
+                (
+                    r.root_hash().clone(),
+                    Retention::Checkpoint {
+                        id: (),
+                        is_marked: false,
+                    },
+                )
+            }),
+        )
+        .map_err(ShardTreeError::Insert)?
+        .expect("slice of inserted roots was verified to be nonempty");
+
+    put_cap(conn, table_prefix, cap_result.subtree.take_root()).map_err(ShardTreeError::Storage)?;
```

Comment:
`subtree.take_root()` is a confusing method name, as it suggests this takes the root node out of the cap tree leaving `subtree` usable, when in fact it consumes `self` and returns the entire cap tree structure (so that we can later reconstruct it using `LocatedTree::from_parts` above. I expect it would be more accurately named `into_tree` or `into_inner` or `into_unlocated`.

We should do a full pass over the `subtree` crate to fix this and similar non-standard usages of the `take_` prefix.

Reply:
Yeah, I keep forgetting about the `into` convention.